vncviewer/win32TouchHandler.cxx

```diff
@@ -0,0 +1,251 @@
+
+  case GID_TWOFINGERTAP:
+    ghev.gesture = GH_TWOTAP;
+    break;
+
+  }
+
+  dbgPrint(gi.dwID, ghev);
+  BaseTouchHandler::handleGestureEvent(ghev);
+}
+
+void win32TouchHandler::fakeMotionEvent(const GHEvent origEvent) {
+  UINT Msg = WM_MOUSEMOVE;
+  WPARAM wParam = MAKEWPARAM(buttonmask, 0);
+  LPARAM lParam = MAKELPARAM(origEvent.event_x, origEvent.event_y);
+
+  dbgPrint(Msg, origEvent);
+  pushFakeEvent(Msg, wParam, lParam);
+}
+
+void win32TouchHandler::fakeButtonEvent(bool press, int button,
+                                        const GHEvent origEvent) {
+  UINT Msg = 0;
+  WPARAM wParam = MAKEWPARAM(0, 0);
+  LPARAM lParam = MAKELPARAM(0, 0);
+  int delta = 0;
+
+  switch (button) {
+
+  case 1: // left mousebutton
+    if (press) {
+      Msg = WM_LBUTTONDOWN;
+      buttonmask |= MK_LBUTTON;
+    } else {
+      Msg = WM_LBUTTONUP;
+      buttonmask &= ~MK_LBUTTON;
+    }
+    break;
+  case 2: // middle mousebutton
+    if (press) {
+      Msg = WM_MBUTTONDOWN;
+      buttonmask |= MK_MBUTTON;
+    } else {
+      Msg = WM_MBUTTONUP;
+      buttonmask &= ~MK_MBUTTON;
+    }
+    break;
+  case 3: // right mousebutton
+    if (press) {
+      Msg = WM_RBUTTONDOWN;
+      buttonmask |= MK_RBUTTON;
+    } else {
+      Msg = WM_RBUTTONUP;
+      buttonmask &= ~MK_RBUTTON;
+    }
+    break;
+
+  case 4: // scroll up
+    Msg = WM_MOUSEWHEEL;
+    delta = WHEEL_DELTA;
+    break;
+  case 5: // scroll down
+    Msg = WM_MOUSEWHEEL;
+    delta = -WHEEL_DELTA;
+    break;
+  case 6: // scroll left
+    Msg = WM_MOUSEHWHEEL;
+    delta = -WHEEL_DELTA;
+    break;
+  case 7: // scroll right
+    Msg = WM_MOUSEHWHEEL;
+    delta = WHEEL_DELTA;
+    break;
+
+  default:
+    vlog.error(_("Invalid mouse button %i, must be a number between 1 and 7."),
+               button);
+    return;
+  }
+
+  if (1 <= button && button <= 3) {
+    wParam = MAKEWPARAM(buttonmask, 0);
+
+    // Regular mouse events expect client coordinates
+    lParam = MAKELPARAM(origEvent.event_x, origEvent.event_y);
+
+  } else if (4 <= button && button <= 7) {
+    // Only act on wheel press, not on release
+    if (!press)
+      return;
+
+    wParam = MAKEWPARAM(buttonmask, delta);
+
+    // Wheel events require screen coordinates
+    POINT pos[] = { (LONG)origEvent.event_x, (LONG)origEvent.event_y };
+    ClientToScreen(hWnd, pos);
+    lParam = MAKELPARAM(pos->x, pos->y);
+  }
+
+  dbgPrint(Msg, origEvent);
+  pushFakeEvent(Msg, wParam, lParam);
+}
+
+void win32TouchHandler::fakeKeyEvent(bool press, int keysym,
+                                     const GHEvent origEvent) {
+  UINT Msg = press ? WM_KEYDOWN : WM_KEYUP;
+  WPARAM wParam = MAKEWPARAM(0, 0);
+  LPARAM lParam = MAKEWPARAM(0, 0);
+  int vKey = 0x00;
+  int scancode = 0x00;
+
+  switch(keysym) {
+
+  case XK_Control_L:
+    vKey = VK_CONTROL;
+    scancode = 0x1d;
+    if (press)
+      buttonmask |= MK_CONTROL;
+    else
+      buttonmask &= ~MK_CONTROL;
+    break;
+
+  case XK_Shift_L:
```

Comment:
Was this for testing?

Reply:
Yes, but perhaps we keep it anyway since there's a mask bit for it in windows?

Reply:
Sure. Perhaps with a comment...

Reply:
Fixed in 1d781fd